pstoreds/addr_book.go

```diff
@@ -240,7 +255,91 @@ func (ab *dsAddrBook) AddAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duratio
 		return
 	}
 	addrs = cleanAddrs(addrs)
-	ab.setAddrs(p, addrs, ttl, ttlExtend)
+	ab.setAddrs(p, addrs, ttl, ttlExtend, false)
+}
+
+// ConsumePeerRecord adds addresses from a signed peer.PeerRecord (contained in
+// a record.Envelope), which will expire after the given TTL.
+// See https://godoc.org/github.com/libp2p/go-libp2p-core/peerstore#CertifiedAddrBook for more details.
+func (ab *dsAddrBook) ConsumePeerRecord(recordEnvelope *record.Envelope, ttl time.Duration) (bool, error) {
+	r, err := recordEnvelope.Record()
+	if err != nil {
+		return false, err
+	}
+	rec, ok := r.(*peer.PeerRecord)
+	if !ok {
+		return false, fmt.Errorf("envelope did not contain PeerRecord")
+	}
+	if !rec.PeerID.MatchesPublicKey(recordEnvelope.PublicKey) {
+		return false, fmt.Errorf("signing key does not match PeerID in PeerRecord")
+	}
+
+	// ensure that the seq number from envelope is > any previously received seq no
+	if ab.latestPeerRecordSeq(rec.PeerID) >= rec.Seq {
+		return false, nil
+	}
+
+	addrs := cleanAddrs(rec.Addrs)
+	err = ab.setAddrs(rec.PeerID, addrs, ttl, ttlExtend, true)
+	if err != nil {
+		return false, err
+	}
+
+	err = ab.storeSignedPeerRecord(rec.PeerID, recordEnvelope, rec)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (ab *dsAddrBook) latestPeerRecordSeq(p peer.ID) uint64 {
+	pr, err := ab.loadRecord(p, true, false)
+	if err != nil || len(pr.Addrs) == 0 || pr.CertifiedRecord == nil || len(pr.CertifiedRecord.Raw) == 0 {
+		return 0
+	}
+	return pr.CertifiedRecord.Seq
+}
+
+func (ab *dsAddrBook) storeSignedPeerRecord(p peer.ID, envelope *record.Envelope, rec *peer.PeerRecord) error {
+	envelopeBytes, err := envelope.Marshal()
+	if err != nil {
+		return err
+	}
+	// reload record and add routing state
+	// this has to be done after we add the addresses, since if
+	// we try to flush a datastore record with no addresses,
+	// it will just get deleted
+	pr, err := ab.loadRecord(p, true, false)
+	if err != nil {
+		return err
+	}
+	pr.CertifiedRecord = &pb.AddrBookRecord_CertifiedRecord{
+		Seq: rec.Seq,
+		Raw: envelopeBytes,
+	}
+	pr.dirty = true
+	err = pr.flush(ab.ds)
+	return err
+}
+
+// GetPeerRecord returns a record.Envelope containing a peer.PeerRecord for the
+// given peer id, if one exists.
+// Returns nil if no signed PeerRecord exists for the peer.
+func (ab *dsAddrBook) GetPeerRecord(p peer.ID) *record.Envelope {
```

Comment:
Hmmmm... I think this this method should return `(*record.Envelope, ok bool, err error)`. Rationale:

* the `ok` bool is idiomatic in go.
* we are currently swallowing errors.